chess-AI/figshare_api.py

```diff
@@ -0,0 +1,174 @@
+"""Module for retrieving and uploading articles through the Figshare API."""
+import hashlib
+import json
+import os
+
+from urllib.request import urlretrieve
+import requests
+from requests.exceptions import HTTPError
+
+class FigshareApi:
+    """Class with all the functions for interacting with Figshare."""
+    # Size of upload chunks
+    CHUNK_SIZE = 1048576
+    # Figshare base URL
+    BASE_URL = 'https://api.figshare.com/v2/{endpoint}'
+
+    def __init__(self):
+        # API key needs to be stored as an environment variable
+        self.api_key = os.getenv('FIGSHARE_KEY')
+
+    def get_figshare_article(self, store_dir, file_name):
+        """Get an article from figshare and store it in store_path.
+
+        file_name refers to the figshare filename, not title.
+        """
+        endpoint = "account/articles/{article_id}/files"
+        download_url = None
+        store_path = store_dir + file_name
+        # Get all the articles within figshare
+        articles = self.get_articles()
+        # Iterate through all the returned articles
+        for article in articles:
+            article_id = article['id']
+            file_data = self.issue_request('GET', endpoint.format(article_id=article_id))
+            # Ensure articles has associated file data
+            if len(file_data) > 0:
+                # Get the file name of the article
+                figshare_file = file_data[0]['name']
+                # See if this the file being searched for
+                if figshare_file == file_name:
+                    # Get the download URL
+                    download_url = file_data[0]["download_url"]
+                    break
+        if not download_url:
+            print("File %s not found in Figshare" % (file_name))
+            return False
+
+        urlretrieve(download_url, store_path)
+        print("File %s downloaded to %s" % (file_name, store_path))
+        # Return true to signify file was found
+        return True
+
+    def raw_issue_request(self, method, url, data=None, binary=False):
+        """Utility function taken from Figshare documentation."""
+        headers = {'Authorization': 'token ' + self.api_key}
+
+        if data is not None and not binary:
+            data = json.dumps(data)
+
+        response = requests.request(method, url, headers=headers, data=data)
+        try:
+            response.raise_for_status()
+            try:
+                data = json.loads(response.content)
+            except ValueError:
+                data = response.content
+        except HTTPError as error:
+            print("Caught an HTTPError: {}".format(error.message))
+            print('Body:\n', response.content)
+            raise
```

Comment:
I'm not sure if there is something missing here, should there be a `raise ValueError` or something?

Reply:
Not sure, that was taken straight out of Figshare docs